Add subdir property namer

psyray opened the pull request:
Hi,

Is there a way to implement something like this PR ?
Inspired from the property namer, but for directory.
Tested on PHP 7.1.1, Symfony 3.2.9.

I tried to run the test suite, but I can't install mongodb in my php environment (facing problem that I couldn't solve quickly).

So someone could do this for me ?

Thanks

Koc commented:
You can test using travis. Also I think better add option for current subdir namer

Koc commented on Naming/SubdirDirectoryPropertyNamer.php:
```diff
@@ -0,0 +1,63 @@
+        return $name;
+    }
+
+    private function getPropertyValue($object, $propertyPath)
```
Please reuse already configured property accessor from container instead creating new one

psyray replied:
You mean from the one from PropertyNamer ? or the one from PropertyMapping ?
I think you talk about the method readProperty from PropertyMapping.
So, the one from PropertyNamer should be change too, isn't it ?

Koc replied:
Currently `readProperty` supports only properties wich configured for file mapping. Take look at form type

1. https://github.com/dustin10/VichUploaderBundle/blob/master/Form/Type/VichFileType.php#L54
2. https://github.com/dustin10/VichUploaderBundle/blob/master/Resources/config/form.xml#L11

It gets configured by Symfony instance of `PropertyAccessor`.

psyray replied:
I tried to use the **readProperty** method through the **$mapping** parameter of the **directoryName** method but it raise an error because of the **array_key_exists** check
https://github.com/dustin10/VichUploaderBundle/blob/56c0d3ad6fc7c92bae7b05e46b3b9b5be0ec5ecc/Mapping/PropertyMapping.php#L137-L139
The property/method to access (the one added in the config.yml) is not in the **propertyPaths** array.
Any idea ?

psyray replied:
Done, so the same should be applied to the PropertyName class
https://github.com/dustin10/VichUploaderBundle/blob/56c0d3ad6fc7c92bae7b05e46b3b9b5be0ec5ecc/Naming/PropertyNamer.php

garak commented on Naming/SubdirDirectoryPropertyNamer.php:
```diff
@@ -0,0 +1,63 @@
+use Vich\UploaderBundle\Mapping\PropertyMapping;
+use Symfony\Component\PropertyAccess\Exception\NoSuchPropertyException;
+use Vich\UploaderBundle\Exception\NameGenerationException;
+use Symfony\Component\PropertyAccess\PropertyAccess;
```
Can you sort here?

psyray replied:
Done

garak replied:
Not done.
Can you please use php-cs-fixer, as suggested on our contributing file?

psyray replied:
Done (you have some misordering in the storage folder too ;) )

psyray replied:
What's the importance of sorting use declaration ? (quickly search on the web but can't find the information)

garak replied:
I'm not sure, maybe avoiding going crazy by reading that list

psyray replied:
That's a good reason ;)

Koc commented:
Much better, need cover with new tests. Ald Imho better rename `SubdirDirectoryPropertyNamer`  => `PropertyDirectoryNamer `

psyray commented:
Class renamed.
How could I test using Travis ? I've never used it.

garak commented:
Every PR is automatically tested by travis.
Anyway, you can run test yourself by installing vendors with `--ignore-platform-reqs`

psyray commented:
Hi,
Why this PR not merged ?

garak commented:
>Why does this PR not merged ?

Because tests are missing

psyray commented:
Tests added, result below.
How can I fix the CSRF protection error ?
```
PHPUnit 6.2.3 by Sebastian Bergmann and contributors.

Testing VichUploaderBundle Test Suite
................................................S..........EE..  63 / 230 ( 27%)
............................................................... 126 / 230 ( 54%)
............................................................... 189 / 230 ( 82%)
.........................................                       230 / 230 (100%)

Time: 616 ms, Memory: 24.00MB

There were 2 errors:

1) Vich\UploaderBundle\Tests\Functional\UploadTest::testFileIsUploadedWithFileType
Symfony\Component\DependencyInjection\Exception\LogicException: CSRF support cannot be enabled as the Security CSRF component is not installed.

/opt/data/Eclipse/git/VichUploaderBundle/vendor/symfony/framework-bundle/DependencyInjection/FrameworkExtension.php:1261
/opt/data/Eclipse/git/VichUploaderBundle/vendor/symfony/framework-bundle/DependencyInjection/FrameworkExtension.php:194
/opt/data/Eclipse/git/VichUploaderBundle/vendor/symfony/dependency-injection/Compiler/MergeExtensionConfigurationPass.php:66
/opt/data/Eclipse/git/VichUploaderBundle/vendor/symfony/http-kernel/DependencyInjection/MergeExtensionConfigurationPass.php:39
/opt/data/Eclipse/git/VichUploaderBundle/vendor/symfony/dependency-injection/Compiler/Compiler.php:143
/opt/data/Eclipse/git/VichUploaderBundle/vendor/symfony/dependency-injection/ContainerBuilder.php:746
/opt/data/Eclipse/git/VichUploaderBundle/vendor/symfony/http-kernel/Kernel.php:575
/opt/data/Eclipse/git/VichUploaderBundle/vendor/symfony/http-kernel/Kernel.php:117
/opt/data/Eclipse/git/VichUploaderBundle/vendor/symfony/framework-bundle/Test/KernelTestCase.php:158
/opt/data/Eclipse/git/VichUploaderBundle/vendor/symfony/framework-bundle/Test/WebTestCase.php:33
/opt/data/Eclipse/git/VichUploaderBundle/Tests/Functional/UploadTest.php:9

2) Vich\UploaderBundle\Tests\Functional\UploadTest::testFileIsUploadedWithImageType
Symfony\Component\DependencyInjection\Exception\LogicException: CSRF support cannot be enabled as the Security CSRF component is not installed.

/opt/data/Eclipse/git/VichUploaderBundle/vendor/symfony/framework-bundle/DependencyInjection/FrameworkExtension.php:1261
/opt/data/Eclipse/git/VichUploaderBundle/vendor/symfony/framework-bundle/DependencyInjection/FrameworkExtension.php:194
/opt/data/Eclipse/git/VichUploaderBundle/vendor/symfony/dependency-injection/Compiler/MergeExtensionConfigurationPass.php:66
/opt/data/Eclipse/git/VichUploaderBundle/vendor/symfony/http-kernel/DependencyInjection/MergeExtensionConfigurationPass.php:39
/opt/data/Eclipse/git/VichUploaderBundle/vendor/symfony/dependency-injection/Compiler/Compiler.php:143
/opt/data/Eclipse/git/VichUploaderBundle/vendor/symfony/dependency-injection/ContainerBuilder.php:746
/opt/data/Eclipse/git/VichUploaderBundle/vendor/symfony/http-kernel/Kernel.php:575
/opt/data/Eclipse/git/VichUploaderBundle/vendor/symfony/http-kernel/Kernel.php:117
/opt/data/Eclipse/git/VichUploaderBundle/vendor/symfony/framework-bundle/Test/KernelTestCase.php:158
/opt/data/Eclipse/git/VichUploaderBundle/vendor/symfony/framework-bundle/Test/WebTestCase.php:33
/opt/data/Eclipse/git/VichUploaderBundle/Tests/Functional/UploadTest.php:47

ERRORS!
Tests: 230, Assertions: 513, Errors: 2, Skipped: 1.
```
I also added the transliterate option, renamed the service according to the previous asked class rename & updated the doc.
All has been tested on my project, and working fine.

Waiting remarks.

psyray commented:
Any news ?

garak commented:
Please solve pending reviews

psyray commented:
You talk about the CSRF Protection ?
Don't know what's the problem is, this errors cames when I pull changes to rebase my PR.
Any Idea ?
I run the composer update with the `--ignore-platform-reqs` arg

garak commented:
https://help.github.com/articles/about-pull-request-reviews/

psyray commented:
I'm sorry but I did not see any pending reviews...

garak commented:
You should see one here https://github.com/dustin10/VichUploaderBundle/pull/718/files#diff-7f3ffa5ded2a6aa0db0d7773f4ee6b9dR24

psyray commented:
I think I guess the problem, there is too much space after comma, right ?
But why I couldn't see a review with the error ?
I've read the link you posted above, no way to find a review or an error in the conversation view.
Even if I use the `review-requested:USERNAME` in the PR search bar, no review.
So I'm OK to solve my PR but I need some qualified feedback.

garak commented:
Please run php-cs-fixer on your branch

psyray commented:
Result of the latest php-cs-fixer run on my branch
```
If you need help while solving warnings, ask at https://gitter.im/PHP-CS-Fixer, we will help you!
Loaded config default from "/opt/data/Eclipse/git/VichUploaderBundle/.php_cs".
Using cache file ".php_cs.cache".
Paths from configuration file have been overridden by paths provided as command arguments.
   1) /opt/data/Eclipse/git/VichUploaderBundle//Handler/AbstractHandler.php
   2) /opt/data/Eclipse/git/VichUploaderBundle//Handler/UploadHandler.php
   3) /opt/data/Eclipse/git/VichUploaderBundle//Storage/FlysystemStorage.php
   4) /opt/data/Eclipse/git/VichUploaderBundle//Storage/GaufretteStorage.php
   5) /opt/data/Eclipse/git/VichUploaderBundle//Storage/AbstractStorage.php
   6) /opt/data/Eclipse/git/VichUploaderBundle//Metadata/Driver/AnnotationDriver.php
   7) /opt/data/Eclipse/git/VichUploaderBundle//Metadata/Driver/YamlDriver.php
   8) /opt/data/Eclipse/git/VichUploaderBundle//Metadata/Driver/XmlDriver.php
   9) /opt/data/Eclipse/git/VichUploaderBundle//DependencyInjection/Compiler/RegisterPropelModelsPass.php
  10) /opt/data/Eclipse/git/VichUploaderBundle//DependencyInjection/VichUploaderExtension.php
  11) /opt/data/Eclipse/git/VichUploaderBundle//DependencyInjection/Configuration.php
  12) /opt/data/Eclipse/git/VichUploaderBundle//Injector/FileInjector.php
  13) /opt/data/Eclipse/git/VichUploaderBundle//Mapping/PropertyMappingFactory.php
  14) /opt/data/Eclipse/git/VichUploaderBundle//Mapping/PropertyMapping.php
  15) /opt/data/Eclipse/git/VichUploaderBundle//Tests/Fixtures/App/src/TestBundle/Controller/DefaultController.php

Fixed all files in 3.255 seconds, 10.000 MB memory used
```
Seems there is lot of files which was modified but not the one you give me.
So what ?

garak commented:
This is my output:
```
$ php-cs-fixer --version
PHP CS Fixer 2.7.1 Sandy Pool by Fabien Potencier and Dariusz Ruminski
$ php-cs-fixer fix -v --rules=-yoda_style
Loaded config default from "/VichUploaderBundle/.php_cs".
Using cache file ".php_cs.cache".
................................................................
Legend: ?-unknown, I-invalid file syntax, file ignored, S-Skipped, .-no changes, F-fixed, E-error

Checked all files in 0.282 seconds, 8.000 MB memory used
```
(we need to exclude `yoda_style` for now, just because it's a recent rule and it was not previously applied)

psyray commented:
My version is 2.7.0 and my output is
```
$ php-cs-fixer --version
PHP CS Fixer 2.7.0 Sandy Pool by Fabien Potencier and Dariusz Ruminski (e4e93a1)
$ php-cs-fixer fix -v --rules=-yoda_style
Loaded config default from "/VichUploaderBundle/.php_cs".
.................................................................
Legend: ?-unknown, I-invalid file syntax, file ignored, S-Skipped, .-no changes, F-fixed, E-error

Fixed all files in 2.064 seconds, 10.000 MB memory used
```
I'm testing with the 2.7.1 (i'm force update because standard update only give me 2.7.0)

psyray commented:
Result on my add-subdir-property-namer branch, no errors.
Don't know what to do...
```
$ php-cs-fixer --version
PHP CS Fixer 2.7.1 Sandy Pool by Fabien Potencier and Dariusz Ruminski (ab2e189)
$ php-cs-fixer fix -v --rules=-yoda_style
Loaded config default from "/VichUploaderBundle/.php_cs".
.................................................................
Legend: ?-unknown, I-invalid file syntax, file ignored, S-Skipped, .-no changes, F-fixed, E-error

Fixed all files in 2.024 seconds, 10.000 MB memory used
```

garak commented:
Thanks

psyray commented:
:+1:
